skimage/restoration/inpaint.py

```diff
@@ -1,79 +1,184 @@
+import functools
+import operator
 
 import numpy as np
 from scipy import sparse
 from scipy.sparse.linalg import spsolve
 import scipy.ndimage as ndi
 from scipy.ndimage.filters import laplace
+
 import skimage
-from ..measure import label
+from ..measure import label, regionprops_table
+from ._inpaint import _build_matrix_inner
 
 
 def _get_neighborhood(nd_idx, radius, nd_shape):
-    bounds_lo = (nd_idx - radius).clip(min=0)
-    bounds_hi = (nd_idx + radius + 1).clip(max=nd_shape)
+    bounds_lo = tuple(max(p - radius, 0) for p in nd_idx)
+    bounds_hi = tuple(min(p + radius + 1, s) for p, s in zip(nd_idx, nd_shape))
```

Comment:
are the values the same after the change or not? I'm not sure I understand the reason of the change here.

Reply:
The behavior is the same, but it avoids overhead of NumPy calls on short length 2 or 3 arrays. Originally this function was potentially being called for each value in the mask so it could add up over thousands of calls. After the loop reorganization, it is now only being called for mask locations that are near the boundary, so I doubt it is very critical to keep this change. 

Here is a short example:
```Python
radius = 2
nd_idx = np.asarray([64, 83])

%timeit tuple(max(p - radius, 0) for p in nd_idx)
# 1.48 µs ± 26 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)

%timeit (nd_idx - radius).clip(min=0)
# 7.99 µs ± 151 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)
```
so the overhead is only around 6.5 microseconds per call on my machine, but it was being called thousands of times. Similar changes were also realized outside of the function as well since the function now returns tuples instead of small arrays.

Reply:
Thank you for this finding @grlee77, I can confirm how `np.clip` is surprisingly slow, but I found that `np.maximum` do perfectly the job here
```python
In [39]: radius = 2                                                                                      

In [40]: nd_idx = np.asarray([64, 83])                                                                   

In [41]: %timeit tuple(max(p - radius, 0) for p in nd_idx)                                               
2.12 µs ± 13.1 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [42]: %timeit (nd_idx - radius).clip(min=0)                                                           
10.3 µs ± 59.8 ns per loop (mean ± std. dev. of 7 runs, 100000 loops each)

In [43]: %timeit np.maximum(nd_idx - radius, 0)                                                          
1.26 µs ± 2.77 ns per loop (mean ± std. dev. of 7 runs, 1000000 loops each)
```